### DISCUSSION TITLE
Any testers out there? [aomenc improvements]

### DISCUSSION THREAD

**Discussion by Dendraspis** [staxrip/staxrip#501] [Show and tell]
Hey @44vince44 and @JJKylee ,
hello everybody,

are you interested in testing a new feature? 😉

We have still that feature request open, where @JJKylee asked for the AV1 chunk feature (#368) we know from x265 and the request for a decoder selection (#497).

Here will be some improvements I would like you to test and comment. ~~Don't expect them to be bullet proof at the moment.~~ 😁

#### At the moment all improvements are for ***testing purposes only***! 

- [StaxRip v2.1.8.0D_Aomenc1.zip](https://github.com/staxrip/staxrip/files/5952771/StaxRip.v2.1.8.0D_Aomenc1.zip)
  - Initial
- [StaxRip v2.1.8.0D_Aomenc2.zip](https://github.com/staxrip/staxrip/files/5958053/StaxRip.v2.1.8.0D_Aomenc2.zip)
  - Rearranged options
  - Small improvements
- [StaxRip v2.1.8.0D_Aomenc3.zip](https://github.com/staxrip/staxrip/files/5958165/StaxRip.v2.1.8.0D_Aomenc3.zip)
  - `--kf-max-dist` default set to 120
  - `--kf-min-dist` default set to 0
  - `--verbose` set as default
- [StaxRip v2.1.8.0D_Aomenc4.zip](https://github.com/staxrip/staxrip/files/5960104/StaxRip.v2.1.8.0D_Aomenc4.zip)
  - UI fixes
- [StaxRip v2.1.8.0D_Aomenc5.zip](https://github.com/staxrip/staxrip/files/5960839/StaxRip.v2.1.8.0D_Aomenc5.zip)
  - Keyframe (default) values adjusted
  - `--fpf` adjusted for chunks
- [StaxRip v2.1.8.0D_Aomenc6.zip](https://github.com/staxrip/staxrip/files/5961405/StaxRip.v2.1.8.0D_Aomenc6.zip)
  - `--fpf` stats file extension changed to `.fpf`
- [StaxRip v2.1.8.0D_Aomenc7.zip](https://github.com/staxrip/staxrip/files/5964075/StaxRip.v2.1.8.0D_Aomenc7.zip)
  - `--webm` parameter fixed
  - `--ivf` set as default
  - Force mkvmerge to write fps in case of aomenc not using `--ivf`

**Comment by Dendraspis** [comment]
@44vince44 
Do you receive notifications for discussions meanwhile?  🤔

**Reply by ghost** [reply]
@Dendraspis Yes I did receive them, but my phone's notifications were muted (probably by accident). But I'm getting them on the app now! Edit: actually the notifications were not muted on the phone. But I got none. However they appear in the app... Restarting the stupid phone...

**Reply by Dendraspis** [reply]
Didn't want to rush, just test if you can see discussions.  😄

**Reply by ghost** [reply]
I'm getting them now, after having restarted the phone!

**Comment by JJKylee** [comment]
Good job, @Dendraspis!
But I haven't had time to update StaxRip yet, so I need to do it first. And other jobs as well.
I guess it'll take several hours. I'll be on testing after that. 😃

**Reply by Dendraspis** [reply]
No hurry.
Btw:  I'm still running v2.1.7.6  😁 

***EDIT:*** How can you say *"Good job"* when you haven't seen anything yet?  😲  Voodoo?

**Reply by JJKylee** [reply]
Just take it as a compliment.
Compliment first, criticism next. 😄

**Reply by Dendraspis** [reply]
Test it first, then decide if it's worth a compliment.  😉
I prefer the opposite direction. 😎

**Reply by ghost** [reply]
I'll do some tests later ! with great pleasure !

**Reply by JJKylee** [reply]
OK. Here's what I think.

1. _avs2pipemod_ works well.

2. But IMO, we need to polish up the progress bar texts. They are messy now.

<avs2pipemod pass 1 & 2>

![StaxRip v2 1 8 0D_Aomenc1_avs2pipemod_pass1_20210209](https://user-images.githubusercontent.com/3157518/107459010-1456a600-6b0a-11eb-831a-ea4c6ada79c3.jpg)

![StaxRip v2 1 8 0D_Aomenc1_avs2pipemod_pass2_20210209](https://user-images.githubusercontent.com/3157518/107459048-2a646680-6b0a-11eb-8a93-12c2733605e8.jpg)

3. IMHO, as for ffmpeg decoder, adopting `-hwaccel` dxva2 or qsv for CPU encoders like `aomenc` and `x265` is not appropriate. It may cause **_bottleneck_** in data transfer b/w the graphics card and motherboard.

![StaxRip v2 1 8 0D_Aomenc1_Decoder_ffmpeg(DXVA2)_pass1_20210209](https://user-images.githubusercontent.com/3157518/107459378-e6be2c80-6b0a-11eb-86a2-f7852b38dc96.jpg)

4. And about `-threads 1`: why do we restrict it to 1?

5. Is `-pix_fmt yuv420p` automatically generated based on the source? Or hard-wired in the code?

6. Chunk encoding seems to work well. But as we learned in the x265 case, I guess it's better to disable it when ffmpeg is chosen as the decoder, considering trims and other complicated situations. (But again, it's up to you.)

**Reply by Dendraspis** [reply]
1.  👍 

2. Yes, awful, but do you know this is not related to my modifications? I ran the command via console and it looks the same, except the ending. And even the newest build looks the same.  🙄 

3. There should never be something like a bottleneck using GPU for decoding, because the CPU is less involved. But as long as it's not the default setting, let's keep it.  😄 

4. Same as for x265. 😄   I think because it is not necessary, because GPU is doing the hard work, and also to not take power from encoding.

5. Set depending on source - but it's the same logic as used for x265.  😊 

6. Let's take it to the next overhaul. At the moment it works like x265, so there should also be some adjustments made (together). 🤔 

- - - - - - - - -

So despite the broken progress test, which is not caused by the new modifications, everything seems to work as intended.  🥳

**Reply by ghost** [reply]
@Dendraspis  Still have to test the chunk... I' doing some customizations on my 2.1.8.0 and when I'm finished I'll try chunks.

I confirm that all my tests on aomenc on the crashing video that we're having much fun with led to no problem with the source/pipe. So I have nothing to disagree with @JJKylee 😁 

The crashing video's crash, BTW, was related to --kf-max-dist:  instead of using the default 250 (which removes the switch), values of 120, 125 made the video successfully pass. Which BTW  made me spot a small bug (see my new issue 😄 )

**Reply by Dendraspis** [reply]
You can open/make your own answer to separate the tests a bit to get better structure, whereas it's not that important with only 2 testers.  🙄 

But that sounds good - haven't expected almost no issues.  😂 

> The crashing video's crash, BTW, was related to --kf-max-dist: instead of using the default 250 (which removes the switch), values of 120, 125 made the video successfully pass.

Interesting, so it seems to be really a bug in aomenc.  😲

**Reply by JJKylee** [reply]
> 2. Yes, awful, but do you know this is not related to my modifications? I ran the command via console and it looks the same, except the ending. And even the newest build looks the same.  🙄

I see. Didn't know cuz I didn't do it on CMD.

> 3. There should never be something like a bottleneck using GPU for decoding, because the CPU is less involved. But as long as it's not the default setting, let's keep it.  😄

Well, in my experience, the bottleneck _does_ exist when I use GPU decoding -> CPU filters -> GPU encoding(NVEnc). But like you said, it may be insignificantly small in this case since there's no CPU filters involved and aomenc is _very_ slow.

Then if you don't mind, can you add the non-hwaccel version of ffmpeg as another decoder option as well? 

> 4. Same as for x265. 😄   I think because it is not necessary, because GPU is doing the hard work, and also to not take power from encoding.
> 5. Set depending on source - but it's the same logic as used for x265.  😊
> 6. Let's take it to the next overhaul. At the moment it works like x265, so there should also be some adjustments made (together). 🤔

I see.

> So despite the broken progress test, which is not caused by the new modifications, everything seems to work as intended. 🥳

Guess so too. Good job! 👍👍👍

**Reply by Dendraspis** [reply]
2.  Yeah, had to double check it because it looked so weird and I couldn't images this is the original output.  😲 

3. Of course it is, because you use GPU for encoding, so charging it with more tasks like decoding will slow down encoding. ***aomenc*** is CPU based, so using a CPU based decoder is slowing down.
Should be possible. 🤔

**Reply by JJKylee** [reply]
> 2. Of course it is, because you use GPU for encoding, so charging it with more tasks like decoding will slow down encoding. _**aomenc**_ is CPU based, so using a CPU based decoder is slowing down.

AFAIK, CPU and GPU are different. NVENC(encoding) is a different engine than NVDEC(decoding), so retaining the data in the graphics card memory is usually faster than transferring them b/w the graphics card and the motherboard.

![Nvidia_GTX1060_engines_20210210](https://user-images.githubusercontent.com/3157518/107516204-e781ad80-6b60-11eb-8e2c-6c941ac92e20.jpg)

But I agree that a CPU decoder will slow down aomenc.

---

If you encode with NVEnc, you'll see encoding speed is _much faster_ with ffmpeg(DXVA2) or NVENC HARDWARE(NVDEC) as the decoder than with AviSynth/VapourSynth(CPU) as the decoder.

**Reply by Dendraspis** [reply]
> [..] so retaining the data in the graphics card memory is usually faster than transferring them b/w the graphics card and the motherboard.

Even if you GPU decodes the images, the data has to *travel* through the CPU, at least when using ffmpeg. 

> If you encode with NVEnc, you'll see encoding speed is much faster with ffmpeg(DXVA2) or NVENC HARDWARE(NVDEC) as the decoder than with AviSynth/VapourSynth(CPU) as the decoder.

Interesting ... once again something learned.  😊 
Could have something to with the *H.265 encode/​decode* feature of your graphicscard. 🤔

**Reply by JJKylee** [reply]
> Could have something to with the _H.265 encode/​decode_ feature of your graphicscard. 🤔

You're right. The prerequisite is that the graphics card supports hardware decoding for the specific codec.

BTW, `StaxRip.v2.1.8.0D_Aomenc3.zip` looks good to me _except_ that ffmpeg(CPU) decoding is not included.
Don't blame me. I just wanna have a broader options set. 😝

**Reply by Dendraspis** [reply]
>  looks good to me except that ffmpeg(CPU) decoding is not included. Don't blame me.

Two sentences, two complete different meanings. 😂 
But to be honest:  ***NO***, not even x265 has such a decoder. 🤔

**Reply by JJKylee** [reply]
Well, then, can't help it.
As a matter of fact, I rarely use aomenc. 😝

**Reply by Dendraspis** [reply]
Yeah, and  now @44vince44  demonstrated why we should wait a few years ... again!  😒 

At least testing becomes a bit faster due to chunks feature.  😊

**Reply by ghost** [reply]
I'm beginning testing chunks

**Reply by Dendraspis** [reply]
We've actually made it.  🎉 
Thanks for helping.  😊

**Comment by ghost** [comment]
> StaxRip v2.1.8.0D_Aomenc3.zip
> 
>     --kf-max-dist default set to 120
>     --verbose set as default

Thanks!! gimme a couple of hours please before continuing, I've been working on aomenc all morning 🤣  (that was a DELIGHTFUL morning, just missed 3 passes to be perfect)

**Reply by Dendraspis** [reply]
Damn, you were too fast .... I've updated v3, so please re-download.

>  I've been working on aomenc all morning 🤣 (that was a DELIGHTFUL morning, just missed 3 passes to be perfect)

🤣   truly a compensation for the missing 3-pass mode.  😁

**Reply by ghost** [reply]
>  I've updated v3, so please re-download.

done!!

**Comment by ghost** [comment]
I HAVE JUST BEGAN THE TESTINGS now ! (busy on other checkings)

Please there is a major problem that forbids further testing in two pass mode :
First chunk (i=1) is OK but for i >=2 there is a problem

- 🆗 Destination file name of chunk i : [Filename]_out_chunki.webm"
- ❌ Statistics pass file (--fpf) of chunk i : [Filename].txt  **It should be [Filename]_chunki.txt**

**Reply by Dendraspis** [reply]
...and updated.  😁

**Reply by Dendraspis** [reply]
Glad you've seen it.  🏅 

Looks like we are ready for the next release.  😁

**Reply by ghost** [reply]
Probably, but let's have some more testing to be really sure. First tests show no problem!
I'm almost finished. I have the feeling you've got it right from the first time (except for the stats file that is really minor)

**Reply by Dendraspis** [reply]
Meanwhile I changed the extension of the stats files to `.fpf`, but there is no need to test it.  😁 

> (except for the stats file that is really minor)

Not sure if it's minor when an encode is based on wrong data.  🔬

**Reply by ghost** [reply]
conceptually minor ;-)

> Meanwhile I changed the extension of the stats files to .fpf, but there is no need to test it. 😁

if you say so 🤣

**Reply by ghost** [reply]
I found a small problem in the fps of the resulting file. The frames where the cuts happen are calculated OK, and the expect frames are used, but the final fps is not accurate...  Shouldn't it use the fps of the project (= original file or AVS/VS resulting fps) ?
How is the fps of the resulting file calculated

**EDIT:** the problem happens at the muxing stage. Even with 1 chunk the problem exists, fps is not accurate.
e.g: source file's fps = 25, resulting file fps = 25.056
It should be 25.

**Reply by ghost** [reply]
Apparently no value of fps is specified when muxing....but the pipe and encoder all report 25 exact fps.

**Reply by Dendraspis** [reply]
***WTF***

The frame rate it taken from the script - the same number you see right on the main window in the target section.   🤔

**Reply by ghost** [reply]
would you like to see the log ?

**Reply by Dendraspis** [reply]
sure 

Is it possible you are overwriting an encode from a prior test?

**Reply by ghost** [reply]
https://pastebin.com/uYmTzBix

**Reply by ghost** [reply]
yes it is possible indeed, but how would that affect the result ? all files are overwritten ?

**Reply by ghost** [reply]
If i switch back to x265 encoder, the fps of the final file is exactly 25.
Going back again to aomenc, it's back to 25.056.
NOTE: **the problem only happens when I use AVS Trim**. If I don't trim, fps remains exact.
Let me remind you that the frames and the frames count are exact. Only FPS bad.

**Reply by Dendraspis** [reply]
~~Maybe a AVS bug?~~  What says the target fps text on the main window?
Could you test VPY ?

**Reply by ghost** [reply]
It says 25. It's not an AVS issue. Because I switched to x265 (with AVS Pipe to compare same things together) changing nothing except aomenc (AVS, AVSPipe) -> x265 (AVS, AVSPipe).
The result with x265 is 25.00000
Testing VPY now.

**Reply by ghost** [reply]
same thing with VPY
Final fps = 25.056

**Reply by Dendraspis** [reply]
It has to be aomenc that stores a wrong value or MediaInfo that can't get the right number. 🤔

**Reply by ghost** [reply]
No it's not aomenc fault: the webm video stream created by aomenc says 25.00000:


Video (webm)
--
ID : | 1
Format : | AV1
Format/Info : | AOMedia Video 1
Format profile : | Main@L3.0
Codec ID : | V_AV1
Duration : | 18 s 0 ms
Bit rate : | 85.8 kb/s
Width : | 856 pixels
Height : | 480 pixels
Display aspect ratio : | 16:9
Frame rate mode : | Constant
**Frame rate : | 25.000 FPS**

------

**Video (muxed mkv)**

ID : | 1
-- | --
Format : | AV1
Format/Info : | AOMedia Video 1
Format profile : | Main@L3.0
Codec ID : | V_AV1
Duration : | 18 s 0 ms
Bit rate : | 94.8 kb/s
Width : | 856 pixels
Height : | 480 pixels
Display aspect ratio : | 16:9
Frame rate mode : | Variable
**Frame rate : | 25.056 FPS**

**Reply by Dendraspis** [reply]
Is the file small enough to share it with me ?

**Reply by ghost** [reply]
sure, which file you want ?

**Reply by Dendraspis** [reply]
So MediaInfo or mkvmerge...

**Reply by Dendraspis** [reply]
The encode (mkv)

**Reply by ghost** [reply]
Which file(s) you want me to send ?

**Reply by Dendraspis** [reply]
see above  😄

**Reply by ghost** [reply]
Here it is 😄 
[Timecode sample - 25fps output.zip](https://github.com/staxrip/staxrip/files/5961649/Timecode.sample.-.25fps.output.zip)

**Reply by Dendraspis** [reply]
Maybe because  `Frame rate mode                          : Variable`  ?

**Reply by ghost** [reply]
Well I have never specified anywhere variable ! It has invented it. Wait, I'm seeing what happens when I mux the webm externally to mkv

**Reply by Dendraspis** [reply]
`25.056` seems to be right!

`Duration                                 : 00:00:18.000`
`Frame count                              : 451`

`451 / 18 = 25.055555555555555555555555555556`

**Reply by ghost** [reply]
NO! it isn't.

> 451 / 18 = 25.055555555555555555555555555556


OMG LOL please don't 🤣 
Obviously, the duration should be 18.040
The original file is 25.0000 fps, I selected 451 frames. There is no reason the resulting fps suddenly becomes Variable 25.056.

It's mkvmerge fault. Muxing externally gave also 25.056... I think we need to specify fps while muxing...

I bet mkvmerge reads a rounded value of the duration :-(

You know Moritz ? I'll send you a aomenc webm. It's at 25.0000 fps, duration 18.040. If you drop it in mkvtoolnix and mux it to mkv it suddenly becomes 25.056 and duration 18.0000. OK ?

**Reply by ghost** [reply]
LOOOLLLL  **the problem is aomenc:** the webm generated is 25.000 🆗 BUT 18.000 s ❌ .
It seems aomenc does not store duration correctly. So obviously, mkvtoolnix has contractory data, it choses to trust the duration.
Which is the erroneous one, unfortunately.

**Reply by Dendraspis** [reply]
😂  
Of course it is not, another information is:
`Duration                                 : 00:00:18:01` ... 18seconds plus 1 frame
So the calculation is:
`(451  - 1) / 18 = 25`

> It's mkvmerge fault. Muxing externally gave also 25.056... I think we need to specify fps while muxing...
I bet mkvmerge reads a rounded value of the duration :-(

I got the info from your MKV, so the information is there. Seems like MediaInfo shows the wrong value.  🤔 

> You know Moritz ?

Nope.

**Reply by Dendraspis** [reply]
🤔   What trim is applied? Which start and end frame numbers?

**Reply by ghost** [reply]
Please read above I know the answer!

trim applied clip = clip[100:250] + clip[604:804] + clip[1200:1301]

**Reply by Dendraspis** [reply]
...or MediaInfo can't read the AV1 values properly.  🤔 

But correctly it should be:
`Duration : 00:00:18.040`

The question is,  aomenc really stores the duration or only fps and frame count ?!  🤔

**Reply by ghost** [reply]
It seems aomenc store the fps, the frame count, and the rounded value of duration (no milliseconds). 
Mkvtoolnix gives priority to the duration, EXCEPT if you explicitly give him the fps.

With this webm file, muxing to mkv with the parameter "--default-duration 0:25fps" made a PERFECT mkv.

(track 0 , 25fps)

**Reply by Dendraspis** [reply]
Could you try to delete the existing output file and run the encode once more? 
Then the MKV should be okay, too.  🤔

**Reply by ghost** [reply]
Yes I will (although I always check the "Modification Date" 😄  )

Meanwhile, read this: it seems that the duration of the webm created by aomenc is stored with only one decimal: 18s **0** ms

So when I apply the trim, the duration is 451/25 = 18.040 => stored 18.0  we lost 40ms
When I DON'T trim, the duration is 1960/25=78.400 => stored 78.4  we lost NOTHING
This is why the error only appeared when I trimmed.

**Reply by ghost** [reply]
OK I deleted ALL THE TEMP folders, ALL OUTPUTS, RENAMED THE source file, closed staxrip, opened staxrip, dropped the file, set aomenc, Preview -> trim
Encode.
Result : 

18 s 0 ms
--

25.056 FPS
--

--
Is it very hard to add the switch --default-duration [TrackID]:[FPS]fps as a workaround ?

This is important, because when we split into chunks, it is likely that every chunk will "lose" many milliseconds, and when joined and muxed, the resulting file will be a bad variable fps file...

@Dendraspis did you go to sleep ?
Well then good night 😄

**Reply by Dendraspis** [reply]
Seems I was a little bit tired.  🙄   I did, but not on purpose.  😁 

Ok, let's dive into .... 

You gave me the resulting ***MKV*** file `Timecode sample - 25fps output.mkv`.

I tried the following:
`mkvextract '.\Timecode sample - 25fps output.mkv' tracks "0:Timecode sample - 25fps output.webm"`
`mkvmerge '.\Timecode sample - 25fps output.webm' -o '.\Timecode sample - 25fps output_REMERGE.mkv'`

- Timecode sample - 25fps output.mkv
  - `Duration                                 : 18 s 0 ms [General]`
  - `Duration                                 : 18 s 0 ms [Video]`
- Timecode sample - 25fps output.webm
  - `Duration                                 : 18 s 40 ms [General]`
  - `Duration                                 : 18 s 40 ms [Video]`
- Timecode sample - 25fps output_REMERGE.mkv
  - `Duration                                 : 18 s 40 ms [General]`
  - `Duration                                 : 18 s 40 ms [Video]`

The created ***WEBM*** was an IVF container. 
So I reencoded your file without ***IVF*** setting, only ***WEBM*** and the resulting ***WEBM*** showed me:

- Timecode sample - 25fps output_out.webm
  - `Duration                                 : 18 s 0 ms [General]`
  - `Duration                                 : 18 s 0 ms [Video]`

Next try even without ***WEBM*** parameter:

- Timecode sample - 25fps output_out.webm
  - `Duration                                 : 18 s 0 ms [General]`
  - `Duration                                 : 18 s 0 ms [Video]`

Seems like you have to set `--ivf` to create a proper file.  🤔  And I haven't used chunks, trims or anything else - just a clean encode.

I think I'll set `--ivf` as default...

**Reply by ghost** [reply]
HI @Dendraspis .... I was thinking of something like that. Yesterday I made tests with forcing --webm in custom. But it changed nothing. So it confirms what you said above. I'll do now a test forcing --ivf to comfirm also!

**Reply by ghost** [reply]
IT WORKED: even if the intermediate file is *.webm, mediainfo recognizes it as ivf.

Then It is muxed all right:
Mediainfo of the file webm (in reality ivf):

Format : | IVF
-- | --
File size : | 349 KiB
Duration : | 18 s 40 ms
Overall bit rate : | 158 kb/s
FileExtension_Invalid : | ivf


Frame rate : | 25.000 FPS
-- | --


---------------------

MediaInfo of the final file mkv


ID : | 1
-- | --
Format : | AV1
Format/Info : | AOMedia Video 1
Format profile : | Main@L3.1
Codec ID : | V_AV1
Duration : | 18 s 40 ms
Bit rate : | 156 kb/s
Width : | 1 280 pixels
Height : | 720 pixels
Display aspect ratio : | 16:9
Frame rate mode : | Constant
Frame rate : | 25.000 FPS

SO It works. I'll confirm it again with chunks (adding --ivf in the custom switches for both passes)

**Reply by Dendraspis** [reply]
> (adding --ivf in the custom switches for both passes)

The checkbox in the options should be enough. I used 2-pass mode for my encodes above.

**Reply by ghost** [reply]
Well yes, and Value = ivf.
(BTW checkbox of list with 3 options?)

**It worked also with chunks**: here's the mediainfo

Codec ID : | V_AV1
-- | --
Duration : | 18 s 40 ms
Bit rate : | 164 kb/s
Width : | 1 280 pixels
Height : | 720 pixels
Display aspect ratio : | 16:9
Frame rate mode : | Constant
Frame rate : | 25.000 FPS

I checked also the fpf extension of the stats file. EVERYTHING FINE!!!

**Reply by ghost** [reply]
@Dendraspis Above, I show my test results all work Ok with --ivf, including chunk. 
I see you made a selector.
I suggest you remove from` .Switch = "--webm", .Text = "Output WEBM (enabled by default when WebM IO is enabled)` the detail between (), so it will remain `.Switch = "--webm", .Text = "Output WEBM"` because in Staxrip, the user will have to select one of the three options (the default is only valid if no switch is set, which in our case should not be permitted, that is, of course if we use a list with the 3 options)

**Reply by Dendraspis** [reply]
On the other side you can also enable ***WEBM*** and ***IVF*** together like I did.  🤔

**Reply by ghost** [reply]
You mean --webm --ivf in the command line. I didn't know you could do that. Does it have a meaning ? ivf, webm are containers, it's either one or the other, imho ! 
BTW I saw the commits, it should be OK 😄

**Reply by Dendraspis** [reply]
I don't know, but I used both without error.

At the top you can find the new build including the Code Editor fix.  😉

**Reply by Dendraspis** [reply]
Can't believe we've made it.  🤣 Felt like a year. 
Thanks for helping.  😊

**Reply by ghost** [reply]
You did some great work here @Dendraspis ! Credits go to you 👍🏼 👍🏼 👍🏼 😄
